src/joystick/hidapi/SDL_hidapijoystick.c

```diff
@@ -331,31 +420,74 @@ HIDAPI_UpdateDiscovery()
 #endif
 
 #if defined(SDL_USE_LIBUDEV)
-    if (SDL_HIDAPI_discovery.m_nUdevFd >= 0) {
-        /* Drain all notification events.
-         * We don't expect a lot of device notifications so just
-         * do a new discovery on any kind or number of notifications.
-         * This could be made more restrictive if necessary.
-         */
-        for (;;) {
-            struct pollfd PollUdev;
-            struct udev_device *pUdevDevice;
+    if (linux_enumeration_method == ENUMERATION_LIBUDEV) {
+        if (SDL_HIDAPI_discovery.m_nUdevFd >= 0) {
+            /* Drain all notification events.
+             * We don't expect a lot of device notifications so just
+             * do a new discovery on any kind or number of notifications.
+             * This could be made more restrictive if necessary.
+             */
+            for (;;) {
+                struct pollfd PollUdev;
+                struct udev_device *pUdevDevice;
+
+                PollUdev.fd = SDL_HIDAPI_discovery.m_nUdevFd;
+                PollUdev.events = POLLIN;
+                if (poll(&PollUdev, 1, 0) != 1) {
+                    break;
+                }
 
-            PollUdev.fd = SDL_HIDAPI_discovery.m_nUdevFd;
-            PollUdev.events = POLLIN;
-            if (poll(&PollUdev, 1, 0) != 1) {
-                break;
+                SDL_HIDAPI_discovery.m_bHaveDevicesChanged = SDL_TRUE;
+
+                pUdevDevice = usyms->udev_monitor_receive_device(SDL_HIDAPI_discovery.m_pUdevMonitor);
+                if (pUdevDevice) {
+                    usyms->udev_device_unref(pUdevDevice);
+                }
+            }
+        }
+    }
+    else
+#endif /* SDL_USE_LIBUDEV */
+    {
+#if defined(HAVE_INOTIFY)
+        if (inotify_fd >= 0) {
+            union
+            {
+                struct inotify_event event;
+                char storage[4096];
+                char enough_for_inotify[sizeof (struct inotify_event) + NAME_MAX + 1];
+            } buf;
+            ssize_t bytes;
+            size_t remain = 0;
+            size_t len;
+
+            bytes = read(inotify_fd, &buf, sizeof (buf));
+
+            if (bytes > 0) {
+                remain = (size_t) bytes;
             }
 
-            SDL_HIDAPI_discovery.m_bHaveDevicesChanged = SDL_TRUE;
+            while (remain > 0) {
+                if (buf.event.len > 0) {
+                    if (StrHasPrefix(buf.event.name, "hidraw") &&
+                        StrIsInteger(buf.event.name + strlen ("hidraw"))) {
+                        SDL_HIDAPI_discovery.m_bHaveDevicesChanged = SDL_TRUE;
+                        /* We found that an hidraw changed. There is no need
+                         * to continue reading the remaining inotify fd */
```

Comment:
I think we do want to drain the fd here, otherwise the events will hang around indefinitely.

If the fd is polled in a level-triggered way (`select(2)`, `poll(2)`, `epoll(7)` with no special flags), leaving events in the queue means we'll get back here in the next main-loop iteration and repeatedly re-discover HID devices for no good reason, which could be disruptive if it leads to creating and destroying SDL's representation of those devices.

(Or if it's polled in an edge-triggered way (like `epoll(7)` with `EPOLLET`), we'll never get another notification until we have drained the queue.)

Reply:
Ah right. Now I continue to read even if we found an hidraw change, so that we move the read pointer to the end of the fd.